toolchain/lex/tokenized_buffer.cpp

```diff
@@ -355,40 +373,32 @@ auto TokenizedBuffer::SourceBufferDiagnosticConverter::ConvertLoc(
     const char* loc, ContextFnT /*context_fn*/) const -> DiagnosticLoc {
   CARBON_CHECK(StringRefContainsPointer(buffer_->source_->text(), loc))
       << "location not within buffer";
-  int64_t offset = loc - buffer_->source_->text().begin();
+  int32_t offset = loc - buffer_->source_->text().begin();
 
-  // Find the first line starting after the given location. Note that we can't
-  // inspect `line.length` here because it is not necessarily correct for the
-  // final line during lexing (but will be correct later for the parse tree).
-  const auto* line_it = std::partition_point(
+  // Find the first line starting after the given location.
+  const auto* next_line_it = std::partition_point(
       buffer_->line_infos_.begin(), buffer_->line_infos_.end(),
       [offset](const LineInfo& line) { return line.start <= offset; });
 
   // Step back one line to find the line containing the given position.
-  CARBON_CHECK(line_it != buffer_->line_infos_.begin())
+  CARBON_CHECK(next_line_it != buffer_->line_infos_.begin())
       << "location precedes the start of the first line";
-  --line_it;
+  const auto* line_it = std::prev(next_line_it);
   int line_number = line_it - buffer_->line_infos_.begin();
   int column_number = offset - line_it->start;
 
-  // Start by grabbing the line from the buffer. If the line isn't fully lexed,
-  // the length will be npos and the line will be grabbed from the known start
-  // to the end of the buffer; we'll then adjust the length.
-  llvm::StringRef line =
-      buffer_->source_->text().substr(line_it->start, line_it->length);
-  if (line_it->length == static_cast<int32_t>(llvm::StringRef::npos)) {
-    CARBON_CHECK(line.take_front(column_number).count('\n') == 0)
-        << "Currently we assume no unlexed newlines prior to the error column, "
-           "but there was one when erroring at "
-        << buffer_->source_->filename() << ":" << line_number << ":"
-        << column_number;
-    // Look for the next newline since we don't know the length. We can start at
-    // the column because prior newlines will have been lexed.
-    auto end_newline_pos = line.find('\n', column_number);
-    if (end_newline_pos != llvm::StringRef::npos) {
-      line = line.take_front(end_newline_pos);
-    }
-  }
+  // Grab the line from the buffer, either by slicing from this line to the next
+  // minus the newline, or if on the last line from the start to the end of the
+  // buffer.
```

Comment:
Small edit, missing comma:

```suggestion
  // Grab the line from the buffer, either by slicing from this line to the next
  // minus the newline, or if on the last line, from the start to the end of the
  // buffer.
```

Larger edit, splitting out a sentence:

```suggestion
  // Grab the line from the buffer by slicing from this line to the next
  // minus the newline. When on the last line, instead use the start to the end of the
  // buffer.
```